lms/services/roster.py

```diff
@@ -373,21 +504,105 @@ def _get_roster_roles(self, roster) -> list[LTIRole]:
+        ai = self._get_application_instance(lms_course)
+        assert ai.lti_registration, "No LTI registration found for LMSCourse."
+        return ai.lti_registration
+
+    def _get_course_instructor(self, lms_course: LMSCourse) -> LMSUser | None:
+        return self._db.scalars(
+            select(LMSUser)
+            .join(LMSCourseMembership)
+            .join(LTIRole)
+            .where(
+                LMSCourseMembership.lms_course_id == lms_course.id,
+                LTIRole.type == RoleType.INSTRUCTOR,
+                LTIRole.scope == RoleScope.COURSE,
+            )
+            .order_by(LMSCourseMembership.updated.desc())
+        ).first()
+
+    def _get_canvas_sections(
+        self,
+        canvas_api_client: CanvasAPIClient,
+        oauth2_token_service: OAuth2TokenService,
+        lms_course: LMSCourse,
+        with_refresh_token=False,
+    ) -> list[dict]:
+        try:
+            api_sections = canvas_api_client.course_sections(
+                lms_course.lms_api_course_id, with_students=True
+            )
+        except OAuth2TokenError as err:
+            if with_refresh_token and err.refreshable:
+                if not self._refresh_canvas_token(
+                    canvas_api_client, oauth2_token_service
+                ):
+                    LOG.info(
+                        "Failed to fetch sections for course %s, error refreshing token",
+                        lms_course.id,
+                    )
+                    return []
+
+                api_sections = self._get_canvas_sections(
+                    canvas_api_client,
+                    oauth2_token_service,
+                    lms_course,
+                    with_refresh_token=False,
+                )
+            else:
+                LOG.info(
+                    "Failed to fetch sections for course %s, invalid API token",
+                    lms_course.id,
+                )
+                return []
+
+        return api_sections
+
+    def _refresh_canvas_token(
+        self, canvas_service: CanvasAPIClient, oauth2_token_service
+    ) -> bool:
+        try:
+            refresh_token = oauth2_token_service.get().refresh_token
+            canvas_service.get_refreshed_token(refresh_token)
+        except (ConcurrentTokenRefreshError, CanvasAPIError):
+            return False
+
+        return True
+
+    def _get_course_users(self, lms_course: LMSCourse) -> dict[str, LMSUser]:
+        users_from_course_roster = (
+            select(LMSUser)
+            .join(CourseRoster)
+            .where(
+                CourseRoster.lms_course_id == lms_course.id,
+                CourseRoster.active.is_(True),
+            )
+        )
+        users_from_launches = (
+            select(LMSUser)
+            .join(LMSCourseMembership)
+            .where(
+                LMSCourseMembership.lms_course_id == lms_course.id,
+            )
+        )
+        users = self._db.execute(
```

Comment:
Here for completeness, when getting uses from a course we take both the ones from launches and roster.